File: packages/core/src/vbrief-validate/precutover.ts
import { existsSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import {
  resolveLifecycleLayout,
  resolveLifecycleRoot,
  resolveSpecArtifactPath,
} from "../layout/resolve.js";
import {
  contentHasGeneratedSpecSource,
  GENERATED_SPEC_PURPOSE,
} from "../spec-authority/constants.js";
import { isFullSpecState, isGreenfieldSpecExport } from "../spec-authority/resolver.js";
import { DEPRECATION_SENTINEL } from "../vbrief-build/constants.js";

export { DEPRECATION_SENTINEL as DEPRECATED_REDIRECT_SENTINEL };

const DEPRECATION_REDIRECT_PURPOSE = "<!-- Purpose: deprecation redirect -->";

const LIFECYCLE_FOLDERS = ["proposed", "pending", "active", "completed", "cancelled"] as const;

export function missingLifecycleFolders(projectRoot: string): string[] {
  let lifecycleRoot: string;
  try {
    lifecycleRoot = resolveLifecycleRoot(projectRoot);
  } catch {
    // No xbrief/ layout present -- all lifecycle folders are missing.
    return [...LIFECYCLE_FOLDERS];
  }
  return LIFECYCLE_FOLDERS.filter((folder) => !existsSync(join(lifecycleRoot, folder)));
}

function hasCompleteLifecycle(projectRoot: string): boolean {
  return missingLifecycleFolders(projectRoot).length === 0;
}

/** Return true when markdown content is a migration redirect stub. */
export function isDeprecationRedirect(content: string): boolean {
  return content.includes(DEPRECATION_SENTINEL) || content.includes(DEPRECATION_REDIRECT_PURPOSE);
}

/** Full-spec generated export (layout-resolved specification artifact source line). */
export function isGeneratedSpecificationExport(projectRoot: string, content: string): boolean {
  if (!content.includes(GENERATED_SPEC_PURPOSE) || !contentHasGeneratedSpecSource(content)) {
    return false;
  }
  try {
    return existsSync(resolveSpecArtifactPath(projectRoot));
  } catch {
    return false;
  }
}

/** Return true for a fully current generated spec export (full-spec or greenfield). */
export function isCurrentGeneratedSpecification(projectRoot: string, content: string): boolean {
  if (!hasCompleteLifecycle(projectRoot)) return false;
  if (isGeneratedSpecificationExport(projectRoot, content)) return true;
  return isGreenfieldSpecExport(projectRoot);
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function safeReadText(path: string): string {
  try {
    return readFileSync(path, "utf8");
  } catch {
    return "";
  }
}

function rootMarkdownIsLegacy(projectRoot: string, filename: string, content: string): boolean {
  if (isDeprecationRedirect(content)) return false;
  if (filename === "SPECIFICATION.md") {
    if (isGeneratedSpecificationExport(projectRoot, content)) return false;
    if (isGreenfieldSpecExport(projectRoot)) return false;
    if (isFullSpecState(projectRoot)) return false;
  }
  return filename === "SPECIFICATION.md" || filename === "PROJECT.md";
}

/** Return root artifact filenames that are legacy pre-v0.20 inputs (#793 / migrate preflight). */
export function detectPreCutoverLegacy(projectRoot: string): string[] {
  const legacy: string[] = [];
  for (const filename of ["SPECIFICATION.md", "PROJECT.md"] as const) {
    const candidate = join(projectRoot, filename);
    if (!isFile(candidate)) continue;
    const content = safeReadText(candidate);
    if (rootMarkdownIsLegacy(projectRoot, filename, content)) {
      legacy.push(filename);
    }
  }
  return legacy;
}

/** Structured result of a pre-cutover (pre-v0.20 document model) probe. */
export interface PrecutoverDetection {
  preCutover: boolean;
  reasons: string[];
}

export function detectPreCutover(projectRoot: string): PrecutoverDetection {
  const reasons: string[] = [];

  const specPath = join(projectRoot, "SPECIFICATION.md");
  if (isFile(specPath)) {
    const content = safeReadText(specPath);
    if (!isDeprecationRedirect(content) && !isCurrentGeneratedSpecification(projectRoot, content)) {
      reasons.push(
        "SPECIFICATION.md is a pre-v0.20 hand-authored doc (not a deprecation redirect or current generated export)",
      );
    }
  }

  const projectMdPath = join(projectRoot, "PROJECT.md");
  if (isFile(projectMdPath)) {
    const content = safeReadText(projectMdPath);
    if (!isDeprecationRedirect(content)) {
      reasons.push("PROJECT.md is a pre-v0.20 hand-authored doc (not a deprecation redirect)");
    }
  }

  // Since #2112, resolveLifecycleLayout throws when no xbrief/ layout is present.
  // Catch that error so the doctor can still diagnose pre-cutover state on unmigrated trees.
  // However, an empty (greenfield) directory with no pre-cutover artifacts is NOT pre-cutover.
  let layout: ReturnType<typeof resolveLifecycleLayout> | null = null;
  try {
    layout = resolveLifecycleLayout(projectRoot);
  } catch {
    // No xbrief/ layout found. Only report a pre-cutover reason if legacy artifacts
    // (SPECIFICATION.md / PROJECT.md) were already detected above; a bare empty directory
    // without any artifacts is greenfield, not pre-cutover.
    if (reasons.length > 0) {
      reasons.push("xbrief/ lifecycle layout not found -- run `deft migrate:xbrief` first");
    }
    return { preCutover: reasons.length > 0, reasons };
  }
  if (existsSync(layout.root)) {
    const missing = missingLifecycleFolders(projectRoot);
    if (missing.length > 0) {
      reasons.push(`${layout.artifactDir}/ is missing lifecycle folder(s): ${missing.join(", ")}`);
    }
  }

  return { preCutover: reasons.length > 0, reasons };
}

/** Last release that ships `scripts/migrate_vbrief.py` on the consumer deposit path (#2068). */
export const FROZEN_PRECUTOVER_MIGRATION_TAG = "v0.59.0";

export function frozenPreCutoverMigrationGuidance(): string {
  return (
    `Current npm releases no longer ship in-product \`task migrate:vbrief\`. Best-effort path (anchored on the ` +
    `${FROZEN_PRECUTOVER_MIGRATION_TAG} git tag, from which GitHub serves a source tarball on demand): pin framework ` +
    `${FROZEN_PRECUTOVER_MIGRATION_TAG}, install Python 3.11+ and uv, run \`task migrate:vbrief\` once from that payload, ` +
    `then upgrade to current npm. This is a two-hop chain: pre-v0.20 flat model -> vBRIEF v0.6 (on ${FROZEN_PRECUTOVER_MIGRATION_TAG}) ` +
    `-> xBRIEF via \`deft migrate:xbrief\` on current npm. If the ${FROZEN_PRECUTOVER_MIGRATION_TAG} payload is unavailable, ` +
    `fall back to a manual fresh start: \`directive init\` a new project on current npm and hand-port your spec content. ` +
    `See UPGRADING.md § Frozen pre-v0.20 document-model migration (#2068).`
  );
}

export function renderPrecutoverLine(projectRoot: string): string {
  const { preCutover, reasons } = detectPreCutover(projectRoot);
  if (!preCutover) {
    return "Pre-cutover: none -- project is on the current vBRIEF document model.";
  }
  const summary = reasons.join("; ").replace(/\r?\n/g, " ");
  return `Pre-cutover: migration needed -- ${summary}. ${frozenPreCutoverMigrationGuidance()}`;
}

// Re-export for classify callers (#2013).
export { isFullSpecState, isGreenfieldSpecExport };

File: scripts/_precutover.py
"""Shared pre-v0.20 document-model detection helpers.

The session-start guard, CLI gate, validator, and migration preflight all need
the same distinction:

* deprecation redirect stubs are migrated/current enough;
* generated ``SPECIFICATION.md`` exports from ``task spec:render`` are not
  hand-authored legacy docs when their source JSON exists, and are fully
  current vBRIEF artifacts when their lifecycle folders also exist;
* hand-authored root docs are legacy pre-cutover inputs.
"""

from __future__ import annotations

from pathlib import Path

LIFECYCLE_FOLDERS: tuple[str, ...] = (
    "proposed",
    "pending",
    "active",
    "completed",
    "cancelled",
)

DEPRECATED_REDIRECT_SENTINEL = "<!-- deft:deprecated-redirect -->"
DEPRECATION_REDIRECT_PURPOSE = "<!-- Purpose: deprecation redirect -->"

GENERATED_SPEC_PURPOSE = "<!-- Purpose: rendered specification -->"
GENERATED_SPEC_SOURCE = "<!-- Source of truth: vbrief/specification.vbrief.json -->"
GENERATED_SPEC_SOURCE_PD = "<!-- Source of truth: vbrief/PROJECT-DEFINITION.vbrief.json -->"
SPEC_SOURCE_RELPATH = Path("vbrief") / "specification.vbrief.json"
PD_SOURCE_RELPATH = Path("vbrief") / "PROJECT-DEFINITION.vbrief.json"


def missing_lifecycle_folders(project_root: Path) -> list[str]:
    """Return missing vBRIEF lifecycle folder names for ``project_root``."""
    vbrief_root = project_root / "vbrief"
    return [folder for folder in LIFECYCLE_FOLDERS if not (vbrief_root / folder).is_dir()]


def has_complete_lifecycle(project_root: Path) -> bool:
    """Return True when every canonical lifecycle folder exists."""
    return not missing_lifecycle_folders(project_root)


def is_deprecation_redirect(content: str) -> bool:
    """Return True when markdown content is a migration redirect stub."""
    return DEPRECATED_REDIRECT_SENTINEL in content or DEPRECATION_REDIRECT_PURPOSE in content


def _read_text_safe(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return ""


def is_greenfield_spec_export(project_root: Path) -> bool:
    """Greenfield PD-sourced generated SPECIFICATION.md (#2013)."""
    spec_md = _read_text_safe(project_root / "SPECIFICATION.md")
    return (
        (project_root / PD_SOURCE_RELPATH).is_file()
        and not (project_root / SPEC_SOURCE_RELPATH).is_file()
        and GENERATED_SPEC_PURPOSE in spec_md
        and GENERATED_SPEC_SOURCE_PD in spec_md
    )


def is_full_spec_state(project_root: Path) -> bool:
    """Full-spec post-cutover state (spec file + matching banner)."""
    spec_md = _read_text_safe(project_root / "SPECIFICATION.md")
    return (
        (project_root / PD_SOURCE_RELPATH).is_file()
        and has_complete_lifecycle(project_root)
        and GENERATED_SPEC_PURPOSE in spec_md
        and GENERATED_SPEC_SOURCE in spec_md
        and (project_root / SPEC_SOURCE_RELPATH).is_file()
    )


def is_generated_specification_export(project_root: Path, content: str) -> bool:
    """Return True for a generated ``task spec:render`` root export.

    The banner alone is not enough: the declared vBRIEF source must also
    exist. Lifecycle completeness is checked separately by
    ``is_current_generated_specification``.
    """
    return (
        GENERATED_SPEC_PURPOSE in content
        and GENERATED_SPEC_SOURCE in content
        and (project_root / SPEC_SOURCE_RELPATH).is_file()
    )


def is_current_generated_specification(project_root: Path, content: str) -> bool:
    """Return True for a fully current generated spec export (full-spec or greenfield)."""
    if not has_complete_lifecycle(project_root):
        return False
    return is_generated_specification_export(project_root, content) or is_greenfield_spec_export(
        project_root
    )


def root_markdown_is_legacy(project_root: Path, filename: str, content: str) -> bool:
    """Return True if a root markdown artifact should trigger migration."""
    if is_deprecation_redirect(content):
        return False
    if filename == "SPECIFICATION.md":
        if is_generated_specification_export(project_root, content):
            return False
        if is_greenfield_spec_export(project_root):
            return False
    return filename in {"SPECIFICATION.md", "PROJECT.md"}


def detect_pre_cutover_legacy(project_root: Path) -> list[str]:
    """Return root artifact filenames that are legacy pre-v0.20 inputs."""
    legacy: list[str] = []
    for filename in ("SPECIFICATION.md", "PROJECT.md"):
        candidate = project_root / filename
        if not candidate.is_file():
            continue
        try:
            content = candidate.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue
        if root_markdown_is_legacy(project_root, filename, content):
            legacy.append(filename)
    return legacy
